pkg/gatewayserver/io/basicstationlns/basicstationlns.go

```diff
@@ -128,20 +128,48 @@ func (s *srv) handleDiscover(c echo.Context) error {
 }
 
 func (s *srv) handleTraffic(c echo.Context) error {
-	uid := c.Param("uid")
-	ids, err := unique.ToGatewayID(uid)
-	if err != nil {
-		return err
-	}
-	ctx, err := unique.WithContext(s.ctx, uid)
-	if err != nil {
-		return err
+	id := c.Param("id")
+	auth := c.Request().Header.Get(echo.HeaderAuthorization)
+	ctx := s.ctx
+	if auth != "" {
+		md := metadata.New(map[string]string{
+			"id":            id,
+			"authorization": fmt.Sprintf("Bearer %s", auth),
+		})
+		if ctxMd, ok := metadata.FromIncomingContext(s.ctx); ok {
+			md = metadata.Join(ctxMd, md)
+		}
+		ctx = metadata.NewIncomingContext(s.ctx, md)
 	}
-	ctx = log.NewContextWithField(s.ctx, "gateway_uid", uid)
+
 	logger := log.FromContext(ctx).WithFields(log.Fields(
 		"endpoint", "traffic",
 		"remote_addr", c.Request().RemoteAddr,
 	))
+
+	ctx, ids, err := s.server.FillGatewayContext(ctx, ttnpb.GatewayIdentifiers{GatewayID: id})
+	if err != nil {
+		logger.WithError(err).Debug("Failed to fill gateway context")
+		return err
+	}
+
+	if auth != "" {
+		if err = rights.RequireGateway(ctx, ids, ttnpb.RIGHT_GATEWAY_LINK); err != nil {
+			logger.WithError(err).Debug("Failed to authorize gateway")
+			return err
+		}
+	} else {
+		ctx = rights.NewContext(ctx, rights.Rights{
```

Comment:
Yeah this is good; if the gateway doesn't provide auth, we assume it.

Note that if GS doesn't allow unregistered unauthenticated gateways to connect, it returns an error already in `FillGatewayContext`.